File: Empty/library/src/test/java/com/gmail/l2t45s7e9/library/IntegrativeTest.java
package com.gmail.l2t45s7e9.library;

import androidx.annotation.NonNull;
import androidx.arch.core.executor.ArchTaskExecutor;
import androidx.arch.core.executor.TaskExecutor;
import com.gmail.l2t45s7e9.java.entity.BirthDateNotificationModel;
import com.gmail.l2t45s7e9.java.entity.Contact;
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsInteractor;
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsRepository;
import com.gmail.l2t45s7e9.java.interactor.ContactDetilsModel;
import com.gmail.l2t45s7e9.java.interactor.CurrentDate;
import com.gmail.l2t45s7e9.java.interactor.NotificationInteractor;
import com.gmail.l2t45s7e9.java.interactor.NotificationRepository;
import com.gmail.l2t45s7e9.library.domain.ContactDetailsViewModel;
import com.gmail.l2t45s7e9.library.interfaces.SchedulersProvider;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;
import java.util.Calendar;
import java.util.GregorianCalendar;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(MockitoJUnitRunner.class)
public class IntegrativeTest {
    @Mock
    CurrentDate dateModel;
    @Mock
    NotificationRepository notificationRepository;
    @Mock
    ContactDetailsRepository contactDetailsRepository;
    @Mock
    SchedulersProvider schedulersProvider;

    private void installDelegate() {
        ArchTaskExecutor.getInstance().setDelegate(new TaskExecutor() {
            @Override
            public void executeOnDiskIO(@NonNull Runnable runnable) {
                runnable.run();
            }

            @Override
            public void postToMainThread(@NonNull Runnable runnable) {
                runnable.run();
            }

            @Override
            public boolean isMainThread() {
                return true;
            }
        });
    }

    @Test
    public void successfully_adding_notification() {
        installDelegate();
        GregorianCalendar addNotificationDate = new GregorianCalendar(2000, 8, 8);
        GregorianCalendar currentDate = new GregorianCalendar(1999, 8, 9);
        GregorianCalendar birthDate = new GregorianCalendar();
        birthDate.set(Calendar.DATE, 8);
        birthDate.set(Calendar.MONTH, Calendar.SEPTEMBER);
        Contact contact = new Contact(
                "0",
                "Ivan",
                "",
                "",
                "",
                "",
                "",
                birthDate,
                0
        );
        Single<Contact> single = Single.fromCallable(() -> contact);
        Mockito.when(contactDetailsRepository.loadDetailsInformation(contact.getId(), contact.getContactColor())).thenReturn(single);
        Mockito.when(dateModel.getCurrentDate()).thenReturn(currentDate);
        Mockito.when(schedulersProvider.io()).thenReturn(Schedulers.trampoline());
        Mockito.when(schedulersProvider.ui()).thenReturn(Schedulers.trampoline());
        ContactDetailsInteractor contactDetailsInteractor = new ContactDetilsModel(contactDetailsRepository);
        NotificationInteractor notificationInteractor = new BirthDateNotificationModel(dateModel, notificationRepository);
        ContactDetailsViewModel contactDetailsViewModel = new ContactDetailsViewModel(contactDetailsInteractor, notificationInteractor, schedulersProvider);
        contactDetailsViewModel.loadContactDetails(contact.getId(), contact.getContactColor());
        contactDetailsViewModel.setNotification();
        Mockito.verify(notificationRepository).setManager(contact, addNotificationDate);
    }

    @Test
    public void successfully_adding_notification_when_birthday_in_this_year_is_not() {
        installDelegate();
        GregorianCalendar addNotificationDate = new GregorianCalendar(1999, Calendar.SEPTEMBER, 8);
        GregorianCalendar currentDate = new GregorianCalendar(1999, Calendar.SEPTEMBER, 7);
        GregorianCalendar birthDate = new GregorianCalendar();
        birthDate.set(Calendar.DATE, 8);
        birthDate.set(Calendar.MONTH, Calendar.SEPTEMBER);
        Contact contact = new Contact(
                "0",
                "Ivan",
                "",
                "",
                "",
                "",
                "",
                birthDate,
                0
        );
        Single<Contact> single = Single.fromCallable(() -> contact);
        Mockito.when(contactDetailsRepository.loadDetailsInformation(contact.getId(), contact.getContactColor())).thenReturn(single);
        Mockito.when(dateModel.getCurrentDate()).thenReturn(currentDate);
        Mockito.when(schedulersProvider.io()).thenReturn(Schedulers.trampoline());
        Mockito.when(schedulersProvider.ui()).thenReturn(Schedulers.trampoline());
        ContactDetailsInteractor contactDetailsInteractor = new ContactDetilsModel(contactDetailsRepository);
        NotificationInteractor notificationInteractor = new BirthDateNotificationModel(dateModel, notificationRepository);
        ContactDetailsViewModel contactDetailsViewModel = new ContactDetailsViewModel(contactDetailsInteractor, notificationInteractor, schedulersProvider);
        contactDetailsViewModel.loadContactDetails(contact.getId(), contact.getContactColor());
        contactDetailsViewModel.setNotification();
        Mockito.verify(notificationRepository).setManager(contact, addNotificationDate);
    }

    @Test
    public void successfully_deleting_notification() {
        installDelegate();
        GregorianCalendar currentDate = new GregorianCalendar();
        currentDate.set(Calendar.YEAR, 1999);
        GregorianCalendar birthDate = new GregorianCalendar();
        birthDate.set(Calendar.DATE, 8);
        birthDate.set(Calendar.MONTH, Calendar.SEPTEMBER);
        Contact contact = new Contact(
                "0",
                "Ivan",
                "",
                "",
                "",
                "",
                "",
                birthDate,
                0
        );
        Single<Contact> single = Single.fromCallable(() -> contact);
        Mockito.when(contactDetailsRepository.loadDetailsInformation(contact.getId(), contact.getContactColor())).thenReturn(single);
        Mockito.when(dateModel.getCurrentDate()).thenReturn(currentDate);
        Mockito.when(schedulersProvider.io()).thenReturn(Schedulers.trampoline());
        Mockito.when(schedulersProvider.ui()).thenReturn(Schedulers.trampoline());
        ContactDetailsInteractor contactDetailsInteractor = new ContactDetilsModel(contactDetailsRepository);
        NotificationInteractor notificationInteractor = new BirthDateNotificationModel(dateModel, notificationRepository);
        ContactDetailsViewModel contactDetailsViewModel = new ContactDetailsViewModel(contactDetailsInteractor, notificationInteractor, schedulersProvider);
        contactDetailsViewModel.loadContactDetails(contact.getId(), contact.getContactColor());
        contactDetailsViewModel.cancelNotification();
        Mockito.verify(notificationRepository).cancelManager(contact);
    }

    @Test
    public void successfully_adding_notification_for_contact_who_was_born_on_29February() {
        installDelegate();
        GregorianCalendar addNotificationDate = new GregorianCalendar(2000, Calendar.FEBRUARY, 29);
        GregorianCalendar currentDate = new GregorianCalendar(1999, Calendar.MARCH, 2);
        GregorianCalendar birthDate = new GregorianCalendar();
        birthDate.set(Calendar.DATE, 29);
        birthDate.set(Calendar.MONTH, Calendar.FEBRUARY);
        Contact contact = new Contact(
                "0",
                "Pavel",
                "",
                "",
                "",
                "",
                "",
                new GregorianCalendar(1996, 1, 29),
                0
        );
        Single<Contact> single = Single.fromCallable(() -> contact);
        Mockito.when(contactDetailsRepository.loadDetailsInformation(contact.getId(), contact.getContactColor())).thenReturn(single);
        Mockito.when(dateModel.getCurrentDate()).thenReturn(currentDate);
        Mockito.when(schedulersProvider.io()).thenReturn(Schedulers.trampoline());
        Mockito.when(schedulersProvider.ui()).thenReturn(Schedulers.trampoline());
        ContactDetailsInteractor contactDetailsInteractor = new ContactDetilsModel(contactDetailsRepository);
        NotificationInteractor notificationInteractor = new BirthDateNotificationModel(dateModel, notificationRepository);
        ContactDetailsViewModel contactDetailsViewModel = new ContactDetailsViewModel(contactDetailsInteractor, notificationInteractor, schedulersProvider);
        contactDetailsViewModel.loadContactDetails(contact.getId(), contact.getContactColor());
        contactDetailsViewModel.setNotification();
        Mockito.verify(notificationRepository).setManager(contact, addNotificationDate);
    }

    @Test
    public void successfully_adding_notification_for_contact_who_was_born_on_29February_of_leap_year() {
        installDelegate();
        GregorianCalendar addNotificationDate = new GregorianCalendar(2004, Calendar.FEBRUARY, 29);
        GregorianCalendar currentDate = new GregorianCalendar(2000, Calendar.MARCH, 1);
        GregorianCalendar birthDate = new GregorianCalendar();
        birthDate.set(Calendar.DATE, 29);
        birthDate.set(Calendar.MONTH, Calendar.FEBRUARY);
        Contact contact = new Contact(
                "0",
                "Pavel",
                "",
                "",
                "",
                "",
                "",
                new GregorianCalendar(1996, 1, 29),
                0
        );
        Single<Contact> single = Single.fromCallable(() -> contact);
        Mockito.when(contactDetailsRepository.loadDetailsInformation(contact.getId(), contact.getContactColor())).thenReturn(single);
        Mockito.when(dateModel.getCurrentDate()).thenReturn(currentDate);
        Mockito.when(schedulersProvider.io()).thenReturn(Schedulers.trampoline());
        Mockito.when(schedulersProvider.ui()).thenReturn(Schedulers.trampoline());
        ContactDetailsInteractor contactDetailsInteractor = new ContactDetilsModel(contactDetailsRepository);
        NotificationInteractor notificationInteractor = new BirthDateNotificationModel(dateModel, notificationRepository);
        ContactDetailsViewModel contactDetailsViewModel = new ContactDetailsViewModel(contactDetailsInteractor, notificationInteractor, schedulersProvider);
        contactDetailsViewModel.loadContactDetails(contact.getId(), contact.getContactColor());
        contactDetailsViewModel.setNotification();
        Mockito.verify(notificationRepository).setManager(contact, addNotificationDate);
    }
}

File: library/src/test/java/com/gmail/l2t45s7e9/library/IntegrativeTest.kt
package com.gmail.l2t45s7e9.library

import android.content.Context
import com.gmail.l2t45s7e9.java.entity.BirthDateNotificationModel
import com.gmail.l2t45s7e9.java.entity.Contact
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsInteractor
import com.gmail.l2t45s7e9.java.interactor.ContactDetailsRepository
import com.gmail.l2t45s7e9.java.interactor.ContactDetilsModel
import com.gmail.l2t45s7e9.java.interactor.CurrentDate
import com.gmail.l2t45s7e9.java.interactor.NotificationInteractor
import com.gmail.l2t45s7e9.java.interactor.NotificationRepository
import com.gmail.l2t45s7e9.library.domain.ContactDetailsViewModel
import com.gmail.l2t45s7e9.library.fragmentsState.ContactDetailsMapper
import com.gmail.l2t45s7e9.library.interfaces.DispatchersProvider
import io.mockk.coEvery
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.test.TestCoroutineDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runBlockingTest
import kotlinx.coroutines.test.setMain
import org.spekframework.spek2.Spek
import org.spekframework.spek2.dsl.Root
import org.spekframework.spek2.style.specification.Suite
import org.spekframework.spek2.style.specification.describe
import java.util.Calendar
import java.util.GregorianCalendar

@ExperimentalCoroutinesApi
object IntegrativeTest : Spek({
    val contactDetailsViewModelFactory = setUpMocks()
    val notificationRepository: NotificationRepository by memoized()
    val dateModel: CurrentDate by memoized()
    describe("Notification tests without leap year") {
        val birthDate = GregorianCalendar()
        birthDate[Calendar.DATE] = 8
        birthDate[Calendar.MONTH] = Calendar.SEPTEMBER
        val contact = Contact(
            "0",
            "Ivan",
            "",
            "",
            "",
            "",
            "",
            birthDate,
            0
        )
        val factory by memoized { this.contactDetailsViewModelFactory() }
        mockDefaults(contact)

        it("Notification added") {
            every { dateModel.currentDate } returns GregorianCalendar(1999, Calendar.SEPTEMBER, 9)
            runBlockingTest {
                factory.loadContactDetails(contact.id, contact.contactColor)
                factory.setNotification()
            }
            verify {
                notificationRepository.setManager(
                    contact,
                    GregorianCalendar(2000, Calendar.SEPTEMBER, 8)
                )
            }
        }

        it("Notification added and BD in this year is not") {
            every { dateModel.currentDate } returns GregorianCalendar(1999, Calendar.SEPTEMBER, 7)
            runBlockingTest {
                factory.loadContactDetails(contact.id, contact.contactColor)
                factory.setNotification()
            }
            verify {
                notificationRepository.setManager(
                    contact,
                    GregorianCalendar(1999, Calendar.SEPTEMBER, 8)
                )
            }
        }

        it("Notification deleted") {
            every { dateModel.currentDate } returns GregorianCalendar(1999, Calendar.SEPTEMBER, 7)
            runBlockingTest {
                factory.loadContactDetails(contact.id, contact.contactColor)
                factory.cancelNotification()
            }
            verify { notificationRepository.cancelManager(contact) }
        }
    }
    describe("Notification test with leap year") {
        val birthDate = GregorianCalendar(1996, Calendar.FEBRUARY, 29)
        val contact = Contact(
            "0",
            "Pavel",
            "",
            "",
            "",
            "",
            "",
            birthDate,
            0
        )

        val factory by memoized { this.contactDetailsViewModelFactory() }
        mockDefaults(contact)

        it("Notification added to 29 February") {
            every { dateModel.currentDate } returns GregorianCalendar(1999, Calendar.MARCH, 2)
            runBlockingTest {
                factory.loadContactDetails(contact.id, contact.contactColor)
                factory.setNotification()
            }
            verify {
                notificationRepository.setManager(
                    contact,
                    GregorianCalendar(2000, Calendar.FEBRUARY, 29)
                )
            }
        }

        it("Notification added to 29 February of leap year") {
            every { dateModel.currentDate } returns GregorianCalendar(2000, Calendar.MARCH, 1)
            runBlockingTest {
                factory.loadContactDetails(contact.id, contact.contactColor)
                factory.setNotification()
            }
            verify {
                notificationRepository.setManager(
                    contact,
                    GregorianCalendar(2004, Calendar.FEBRUARY, 29)
                )
            }
        }
    }
})

@ExperimentalCoroutinesApi
private fun Root.installDispatchers(testDispatcher: TestCoroutineDispatcher) {
    @Suppress("UNUSED_VARIABLE") val dispatchers by memoized {
        object : DispatchersProvider {
            override fun io(): CoroutineDispatcher {
                return testDispatcher
            }

            override fun ui(): CoroutineDispatcher {
                return testDispatcher
            }

            override fun default(): CoroutineDispatcher {
                return testDispatcher
            }
        }
    }
}

@ExperimentalCoroutinesApi
private fun Root.instantTaskExecute() {
    val testDispatcher = TestCoroutineDispatcher()
    installDispatchers(testDispatcher)

    beforeEachTest {
        Dispatchers.setMain(testDispatcher)
    }

    afterEachTest {
        Dispatchers.resetMain()
        testDispatcher.cleanupTestCoroutines()
    }
}

@ExperimentalCoroutinesApi
@Suppress("UNUSED_VARIABLE")
fun Root.setUpMocks(): Suite.() -> ContactDetailsViewModel {
    instantTaskExecute()

    val dateModel by memoized {
        mockk<CurrentDate>(relaxed = true)
    }

    val context by memoized {
        mockk<Context>(relaxed = true)
    }

    val notificationRepository by memoized {
        mockk<NotificationRepository>(relaxed = true)
    }

    val contactDetailsRepository by memoized {
        mockk<ContactDetailsRepository>(relaxed = true)
    }

    val contactDetailsMapper by memoized {
        mockk<ContactDetailsMapper>(relaxed = true)
    }

    return {
        val dispatchers by memoized<DispatchersProvider>()
        val notificationInteractor: NotificationInteractor = BirthDateNotificationModel(
            dateModel,
            notificationRepository
        )
        val contactDetailsInteractor: ContactDetailsInteractor = ContactDetilsModel(
            contactDetailsRepository
        )
        ContactDetailsViewModel(
            contactDetailsInteractor = contactDetailsInteractor,
            notificationInteractor = notificationInteractor,
            contactDetailsMapper = contactDetailsMapper,
            dispatchersProvider = dispatchers
        )
    }
}

private fun Suite.mockDefaults(contact: Contact) {
    val contactDetailsRepository: ContactDetailsRepository by memoized()
    beforeEachTest {
        coEvery {
            contactDetailsRepository.loadDetailsInformation(contact.id, contact.contactColor)
        } returns flow { emit(contact) }
    }
}
